lib/tests/streamlit/runtime/fragment_test.py

```diff
@@ -294,3 +312,303 @@ def my_fragment():
             patched_run_with_active_hash.reset_mock()
             saved_fragment()
             patched_run_with_active_hash.assert_called_with("some_hash")
+
+# TESTS FOR WRITING TO CONTAINERS OUTSIDE AND INSIDE OF FRAGMENT
+
+ELEMENT_PRODUCER = Callable[[], Any]
+APP_FUNCTION = Callable[[ELEMENT_PRODUCER], None]
+
+
+def _run_fragment_writes_to_outside_container_app(
+    element_producer: ELEMENT_PRODUCER,
+) -> None:
+    """App with container outside of fragment."""
+
+    outside_container = st.container()
+
+    @st.experimental_fragment
+    def _some_method():
+        st.write("Hello")
+        # this is forbidden
+        with outside_container:
+            element_producer()
+
+    _some_method()
+
+
+def _run_fragment_writes_to_nested_outside_container_app(
+    element_producer: ELEMENT_PRODUCER,
+) -> None:
+    """App with nested container outside of fragment."""
+    with st.container():
+        outside_container = st.container()
+
+    @st.experimental_fragment
+    def _some_method():
+        st.write("Hello")
+        # this is forbidden
+        with outside_container:
+            element_producer()
+
+    _some_method()
+
+
+def _run_fragment_writes_to_nested_outside_container_app2(
+    element_producer: ELEMENT_PRODUCER,
+) -> None:
+    """App with nested container outside of fragment writing from nested container."""
+    with st.container():
+        outside_container = st.container()
+
+    @st.experimental_fragment
+    def _some_method():
+        st.write("Hello")
+        # this is forbidden
+        with outside_container:
+            with st.container():
+                element_producer()
+
+    _some_method()
+
+
+def _run_fragment_writes_to_nested_outside_container_app3(
+    element_producer: ELEMENT_PRODUCER,
+) -> None:
+    """App with nested container outside of fragment writing from nested container."""
+    with st.container():
+        outside_container = st.container()
+
+    @st.experimental_fragment
+    def _some_method():
+        st.write("Hello")
+        # this is forbidden
+        with st.container():
+            with outside_container:
+                element_producer()
+
+    _some_method()
+
+
+def _run_fragment_writes_to_inside_container_app(
+    element_producer: ELEMENT_PRODUCER,
+) -> None:
+    """App with container inside of fragment."""
+
+    @st.experimental_fragment
+    def _some_method():
+        inside_container = st.container()
+
+        st.write("Hello")
+        # this is forbidden
+        with inside_container:
+            element_producer()
+
+    _some_method()
+
+
+def _run_fragment_writes_to_nested_inside_container_app(
+    element_producer: ELEMENT_PRODUCER,
+) -> None:
+    """App with container inside of fragment."""
+
+    @st.experimental_fragment
+    def _some_method():
+        inside_container = st.container()
+
+        st.write("Hello")
+        # this is forbidden
```

Comment:
nit: incorrect "this is forbidden" comments probably a result of a copy-paste

Reply:
oops, I have moved them now, thank you!